### PR TITLE
fix: Explicitly exit when container is out-of-memory

### PR DESCRIPTION
**Pull request by mildaniel**
#### Which issue(s) does this change fix?

https://github.com/aws/aws-sam-cli/issues/6260

#### Why is this change necessary?
When an invoke container exits because it's out of memory, SAM CLI doesn't communicate this to customers and it's unclear why the invocation failed.

#### How does it address the issue?
Explicitly note the exit reason is out-of-memory and exit with return code 1 instead of 0.

#### What side effects does this change have?
Existing invocations that fail due to out-of-memory will return exit code 1.

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [] Write/update unit tests
- [ ] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

File: samcli/local/docker/container_analyzer.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+        ContainerState:
+            Returns a ContainerState object with relevant container data
+        """
+        if not self.container.id:
+            LOG.debug("Container ID not defined, unable to fetch container state")
+            return ContainerState(DEFAULT_OUT_OF_MEMORY)
+
+        state = self.container_manager.inspect(self.container.id)
+
+        if isinstance(state, bool):
```
Comment (mildaniel) [inline samcli/local/docker/container_analyzer.py:41]: This check is admittedly a little weird, but mypy won't like it if we don't do this, and the reason we return a bool here is to follow the legacy pattern in `container_manager`.

File: samcli/local/docker/manager.py [hunk 2]
```diff
@@ -193,6 +194,25 @@ def has_image(self, image_name):
...
+
+        Returns
+        -------
+        Union[bool, dict]
+            Container inspection state if successful, False otherwise
+        """
+        try:
+            return cast(dict, self.docker_client.api.inspect_container(container))
+        except (docker.errors.APIError, docker.errors.NullResource):
+            return False
```
Comment (mndeveci) [inline samcli/local/docker/manager.py:214]: Let's add some debug logs if we need to investigate an issue later on

File: samcli/local/lambdafn/runtime.py [hunk 3]
```diff
@@ -223,9 +226,30 @@ def _on_invoke_done(self, container):
...
+        Raises
+        -------
+        ContainerFailureError
+            If the exit reason is due to out-of-memory, return exit code 1
+
+        """
+        container_analyzer = ContainerAnalyzer(self._container_manager, container)
+        exit_state = container_analyzer.inspect()
+        if exit_state.out_of_memory:
+            raise ContainerFailureError("Container invocation failed due to maximum memory usage")
```
Comment (mndeveci) [inline samcli/local/lambdafn/runtime.py:251]: Just thinking out loud, do we need to flush any outputs before ending the execution here?

Comment (mildaniel) [reply samcli/local/lambdafn/runtime.py:251]: We should be okay here, the RIE should be done executing and all logs streamed at this point.

**Review by jysheng123 — APPROVED** [review]

**Review by mndeveci — APPROVED** [review]

**Pull request merged**